lib/bap_disasm/bap_disasm_rec.ml

```diff
@@ -241,34 +180,74 @@ let next dis s =
     | [] -> Dis.stop dis s
     | r :: roots when not(Memory.contains s.base r) ->
       loop {s with roots}
-    | r :: roots when Span.mem s.visited r ->
+    | r :: roots when Visited.mem s.visited r ->
       loop {s with roots}
     | addr :: roots ->
-      let mem = match Span.upper_bound s.visited addr with
-        | None -> Memory.view ~from:addr s.base
-        | Some r1 -> Memory.range s.base addr r1  in
-      let mem =
-        Result.map_error mem ~f:(fun err -> Error.tag err "next_root") in
-      mem >>= fun mem -> Dis.jump dis mem {s with roots; addr} in
-  let visited = Span.add s.visited (s.addr, Dis.addr dis) in
-  loop {s with visited}
+      let mem = Memory.view ~from:addr s.base in
+      let mem = Result.map_error mem ~f:(fun err -> Error.tag err "next_root") in
+      mem >>= fun mem ->
+      Dis.jump dis mem {s with roots; addr} in
+  let s = {s with visited = Visited.visit s.visited s.addr } in
+  loop s
 
 let stop_on = [`Valid]
 
+let has_common_destination dests addrs a =
+  let find_destinations a = match Addrs.find dests a with
```

Comment:
This is unnecessary inefficient:
1) you're creating sets every time from a list, when you can store sets in the beginning
2) you're creating sets just to make an intersection, and the use `is_empty` and drop the result. This will create a lot of pressure on the garbage collector. Consider doing `Set.exists xs ~f:(Set.mem ys)` to find if there exists an intersection between two sets.